I2C.discover/1

batate opened the pull request:
I2C sensors can usually be configured with a couple of potential address names. Commercial sensor breakout boards must choose, and they often choose differently. This pull requests allows a sensor library to discover the bus name and device address, making it easier to get started with genservers without specifying options.

fhunleth commented on lib/i2c.ex:
```diff
@@ -217,7 +217,43 @@ defmodule Circuits.I2C do
+  Since I2C devises often have several hard coded addresses 
+  and may be on any bus, it may be helpful to discover the device. 
+  
+  Fails if there is more than one potential match, or there are no matches.
```
I'm torn on failing if there's more than one match since that could happen. It seems like the caller would know whether it wants only the first one or more than that. Plus, this function is a nice one-liner if it returns a list. 

Another option would be to call this `discover_first!` and return the first one found or raise. I haven't seen the code that calls this, but I'm figuring that it's something along the lines of if the user doesn't specify an I2C bus and address that the default is to find one for them.

batate replied:
Sold. I will tweak it and add discover_one!.

fhunleth commented on lib/i2c.ex:
```diff
@@ -217,7 +217,43 @@ defmodule Circuits.I2C do
+    end
+  end
+  
+  @spec discover(binary(), [address()]) :: [{bus(), address()}] | {:error, term()}
```
```suggestion
  @spec discover(bus(), [address()]) :: [{bus(), address()}]
```

batate commented:
I have updated the pull request: 

- I changed `discover/1` to `discover!/1` and made it raise an I2C error if it fails
- I added `discover_one!/1` to safely report on exactly the right device, if only one exists. 

These functions can be used like this: 

```
iex> MySensor.discover()
Found sensor on i2c-1 bus at address 0x77. 
:ok
iex>
```

Or: 

```
iex> MySensor.scan()
Found 2 possibilities
- address 0x76 on i2c-1 bus
- address 0x77 on i2c-1 bus
:ok
iex>
```

Or: 

```
iex> MySensor.quick_start
Started sensor on registered process MySensor on bus i2c-1 at address 0x77. 
:ok
```

fhunleth commented:
The commits in this PR were squashed together and merged as part of #67.
